### PR TITLE
Correctly interpret underlying type specification of enums

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -1160,10 +1160,20 @@ void c_typecheck_baset::typecheck_c_enum_type(typet &type)
...
 
-  if(as_expr.find(ID_enum_underlying_type).is_not_nil())
+  const bool have_underlying_type =
+    type.find_type(ID_enum_underlying_type).is_not_nil();
+
+  if(have_underlying_type)
   {
-    warning().source_location = source_location;
-    warning() << "ignoring specification of underlying type for enum" << eom;
+    typecheck_type(type.add_type(ID_enum_underlying_type));
```
Comment (hannes-steffenhagen-diffblue) [inline src/ansi-c/c_typecheck_type.cpp:1168]: Using `add_something` to refer to something that's already there is pretty oof.

File: regression/cbmc/enum_underlying_type_02/main.c [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include <assert.h>
+
+enum enum1 : unsigned char
+{
+  E1 = 256
```
Comment (hannes-steffenhagen-diffblue) [inline regression/cbmc/enum_underlying_type_02/main.c:5]: Good test case to include!

File: src/ansi-c/c_typecheck_type.cpp [hunk 3]
```diff
@@ -1160,10 +1160,20 @@ void c_typecheck_baset::typecheck_c_enum_type(typet &type)
...
+
+    if(!is_signed_or_unsigned_bitvector(
+         static_cast<const typet &>(type.find(ID_enum_underlying_type))))
+    {
+      error().source_location = source_location;
+      error() << "underlying type for enum must be an integral type" << eom;
+      throw 0;
```
Comment (chrisr-diffblue) [inline src/ansi-c/c_typecheck_type.cpp]: Do we not have a structured exception to throw here? (and in the other cases below)

**Review by chrisr-diffblue — COMMENTED** [review]
Looks good, and thanks for the comprehensive test cases. A couple of minor nitpicks really:
1) Error handling - we put quite a big effort into cleaning up error handling, adding structured exceptions, things like that - so it would be good to avoid the `throw 0` if possible.
2) What error messages does clang emit for the "non-integral underlying type" and "underlying type too narrow for constant" cases? It would be good to make our error messages broadly similar just for ease of users googling for help if they hit them, etc.

File: src/ansi-c/c_typecheck_type.cpp (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/c_typecheck_type.cpp]: Yes this should an `invalid_source_file_exceptiont`

**Comment by hannes-steffenhagen-diffblue** [comment]
@chrisr-diffblue Current errors for these cases in clang look like this:

    test.c:1:10: error: non-integral type 'float' is an invalid underlying type

and

    test.c:1:24: error: enumerator value is not representable in the underlying type 'signed char'

I think the existing error messages we have here are sufficiently close to that. We could include some extra information like what we think the underlying type is and why it doesn't fit and such, but I wouldn't call this a blocker.

**Comment by hannes-steffenhagen-diffblue** [comment]
I'm OK with these changes though I'd also prefer if we didn't proliferate the `throw 0` thing.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

**Comment by chrisr-diffblue** [comment]
> @chrisr-diffblue Current errors for these cases in clang look like this:
> 
> ```
> test.c:1:10: error: non-integral type 'float' is an invalid underlying type
> ```
> 
> and
> 
> ```
> test.c:1:24: error: enumerator value is not representable in the underlying type 'signed char'
> ```
> 
> I think the existing error messages we have here are sufficiently close to that. We could include some extra information like what we think the underlying type is and why it doesn't fit and such, but I wouldn't call this a blocker.

Thanks @hannes-steffenhagen-diffblue - @danpoe If it's not too much of a pain, could you model the error messages on the above? In particular, showing the user the type like that its quite helpful to them.

File: src/ansi-c/c_typecheck_type.cpp (hunk 3 above)
Comment (danpoe) [reply src/ansi-c/c_typecheck_type.cpp]: Done

**Comment by danpoe** [comment]
> > @chrisr-diffblue Current errors for these cases in clang look like this:
> > ```
> > test.c:1:10: error: non-integral type 'float' is an invalid underlying type
> > ```
> > 
> > 
> > and
> > ```
> > test.c:1:24: error: enumerator value is not representable in the underlying type 'signed char'
> > ```
> > 
> > 
> > I think the existing error messages we have here are sufficiently close to that. We could include some extra information like what we think the underlying type is and why it doesn't fit and such, but I wouldn't call this a blocker.
> 
> Thanks @hannes-steffenhagen-diffblue - @danpoe If it's not too much of a pain, could you model the error messages on the above? In particular, showing the user the type like that its quite helpful to them.

Ok, changed the error messages to a text similar to the clang messages and including the type.

**Review by chrisr-diffblue — APPROVED** [review]
Thanks for addressing my nitpicks :-)

**Pull request merged**